File: PokedexObjC/PokedexObjC/PokemonDetailViewController.m
//
//  PokemonDetailViewController.m
//  PokedexObjC
//
//  Created by Sean Acres on 8/16/20.
//  Copyright © 2020 Sean Acres. All rights reserved.
//

#import "PokemonDetailViewController.h"
#import "PokedexObjC-Swift.h"

void *KVOContext = &KVOContext;

@interface PokemonDetailViewController ()

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UILabel *idLabel;
@property (weak, nonatomic) IBOutlet UILabel *abilitiesLabel;

@end

@implementation PokemonDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.pokemon addObserver:self forKeyPath:@"urlString" options:0 context:KVOContext];
    [PokemonController.sharedController fillInDetailsFor:self.pokemon];
}

- (void)updateViews
{
    if (self.pokemon.urlString) {
        self.nameLabel.text = self.pokemon.name;
        self.idLabel.text = self.pokemon.identifier;
        self.abilitiesLabel.text = [self.pokemon.abilities componentsJoinedByString:@", "];
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (context == KVOContext) {
        if ([keyPath isEqualToString:@"urlString"]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self updateViews];
            });
        }
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

- (void)dealloc
{
    [self.pokemon removeObserver:self forKeyPath:@"urlString" context:KVOContext];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: PokedexOBJC/PokedexOBJC/PokemonDetailViewController.swift
//
//  PokemonDetailViewController.swift
//  PokedexOBJC
//
//  Created by denis cedeno on 6/12/20.
//  Copyright © 2020 DenCedeno Co. All rights reserved.
//

import UIKit

class PokemonDetailViewController: UIViewController {
    
    @objc var pokemon: DSCPokemon? {
        didSet {
            updateViews()
        }
    }
    
    var pController: PokemonAPI? {
        didSet {
            updateViews()
        }
    }
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var idLabel: UILabel!
    @IBOutlet weak var ability1Label: UILabel!
    @IBOutlet weak var ability2Label: UILabel!
    @IBOutlet weak var ability3Label: UILabel!
    
    @IBOutlet weak var abilitytextView: UITextView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        updateViews()
            
    }
    
    func updateViews() {
        guard isViewLoaded else { return }
        guard let detailPokemon = pokemon else {
            title = "No Pokemon Found, Sorry buddy"
            nameLabel.text = ""
            idLabel.text = ""
            ability1Label.text = ""
            ability2Label.text = ""
            ability3Label.text = ""
            return
        }
         
            self.title = detailPokemon.name.capitalized
            self.nameLabel.text = detailPokemon.name
            self.idLabel.text = "\(detailPokemon.pokemonID)"
            self.ability1Label.text = ""
            self.abilitytextView.text = "\(detailPokemon.abilities)"
            self.ability2Label.text = ""
            self.ability3Label.text = ""
            updateImage()
    }
    
    func updateImage() {
        guard let detailPokemon = pokemon else { return }
        DispatchQueue.main.async {
            self.imageView.image = detailPokemon.pokemonSprite
        }
    }
    
    
    
}
